### PR TITLE
added APIs to handle github installation and webhook callbacks

### PR DESCRIPTION
**Pull request by MuskanPaliwal**
- manually tested in local env
- Yet to implement the auth_info and the setup repo flow in rust

### Summary by CodeRabbit

- New Feature: Added `installHandler` to handle GitHub installation callbacks. This feature enhances the integration process by handling POST requests, constructing a message payload, and publishing it to a specified topic. On success, users are redirected to `/u`.
- New Feature: Introduced `webhookHandler` to manage GitHub webhook events. This function retrieves repository configurations, constructs a data object, and publishes a message to a pub/sub system. It improves the system's responsiveness to GitHub events and provides appropriate HTTP responses.
- Improvement: Enhanced error handling in both `installHandler` and `webhookHandler` to provide more informative responses and debug logs.

### CODE REVIEW

File: pages/api/github/callbacks/install.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+			res.status(500).json({ error: 'Failed to publish message. Data must be in the form of a Buffer.' });
+		}
+	} else {
+		console.error('TOPIC_NAME not set');
+		res.status(500).json({ error: 'Failed to publish message'});
+	}
+	res.redirect('/u');
+}
+
+export default installHandler;
```
[the remark this answers is not in the document]
Comment (MuskanPaliwal) [reply pages/api/github/callbacks/install.ts:26]: @tapishr what behavior do we expect here? we want user to get redirected to /u right regardless of the success of `publishMessage` function?

File: pages/api/github/callbacks/webhook.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  const msgType = 'webhook_callback';
+  console.debug(`[webookHandler] data = ${JSON.stringify(jsonBody)}`)
+  console.debug(`[webookHandler] topicname = ${topicName}`)
+  console.debug(`[webookHandler] repoConfig = ${JSON.stringify(repoConfig)}`)
+  const topicName_str = topicName as string;
+  const result: string | null = await publishMessage(topicName_str, data, msgType)
+  .catch((error) => {
+    console.error('[webookHandler] Failed to get repoConfig from db :', error);
+    return null;
+  });
```
[the remark this answers is not in the document]
Comment (MuskanPaliwal) [reply pages/api/github/callbacks/webhook.ts:47]: No, I prefer .catch instead try-catch.

File: pages/api/github/callbacks/webhook.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  if (!repoConfig) {
+    res.status(500).json({ error: 'Unable to get repoConfig from db' });
+    return;
+  }
+  const data = {
+    repositoryProvider: 'github',
+    eventPayload: jsonBody,
+    repoConfig: repoConfig
+  };
```
[the remark this answers is not in the document]
Comment (MuskanPaliwal) [reply pages/api/github/callbacks/webhook.ts:37]: let it be.

File: pages/api/github/callbacks/webhook.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  const msgType = 'webhook_callback';
+  console.debug(`[webookHandler] data = ${JSON.stringify(jsonBody)}`)
+  console.debug(`[webookHandler] topicname = ${topicName}`)
+  console.debug(`[webookHandler] repoConfig = ${JSON.stringify(repoConfig)}`)
+  const topicName_str = topicName as string;
```
[the remark this answers is not in the document]
Comment (MuskanPaliwal) [reply pages/api/github/callbacks/webhook.ts:42]: okay

File: pages/api/github/callbacks/webhook.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  const result: string | null = await publishMessage(topicName_str, data, msgType)
+  .catch((error) => {
+    console.error('[webookHandler] Failed to get repoConfig from db :', error);
+    return null;
+  });
+  if (result == null) {
+    res.status(500).json({ error: 'Internal Server Error' });
+    return;
```
[the remark this answers is not in the document]
Comment (MuskanPaliwal) [reply pages/api/github/callbacks/webhook.ts:50]: ohh this a privacy hack. please ignore.

File: pages/api/github/callbacks/webhook.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	});
+	if (!repoConfig) {
+		res.status(500).json({ error: 'Unable to get repoConfig from db' });
+		return;
+	}
+	const data = {
+		repositoryProvider: 'github',
+		eventPayload: jsonBody,
+		repoConfig: repoConfig
+	};
```
[the remark this answers is not in the document]
Comment (MuskanPaliwal) [reply pages/api/github/callbacks/webhook.ts:37]: let it be

File: pages/api/github/callbacks/install.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+	const topicName = process.env.TOPIC_NAME;
+	if (topicName) {
+		const data = {
+			'repository_provider': 'github',
+			'installation_code': req.query.code
+		};
+		const msgType = 'install_callback';
+		console.info("Recieved installation code for github, published to ", topicName);
+		try {
+			await publishMessage(topicName, data, msgType);
```
Comment (tapishr) [inline pages/api/github/callbacks/install.ts:14]: convert this to .catch

File: pages/api/github/callbacks/install.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+			console.error('[install_callback] Failed to publish message:', error);
+			return null;
+		});
+		if (result == null) {
+			res.status(500).json({ error: 'Internal Server Error' });
+			return;
+		}
+	} else {
+		console.error('TOPIC_NAME not set');
+		res.status(500).json({ error: 'Failed to publish message'});
```
Comment (tapishr) [inline pages/api/github/callbacks/install.ts:24]: this will be 400, bad request

File: pages/api/github/callbacks/webhook.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+	const { owner: { username: owner }, name } = jsonBody.repository;
+	const provider = "github";
+	
+	console.info("[webookHandler] Received github webhook event for ", name);
+	const topicName: string | null = await getTopicNameFromDB(owner, name, provider).catch((error) => {
+		console.error('[webhookHandler] Failed to get topic name from db:', error);
+		return null;
+	});
+	if (!topicName) {
+		res.status(500)
```
Comment (tapishr) [inline pages/api/github/callbacks/webhook.ts:17]: change to 400

File: pages/api/github/callbacks/install.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+			console.error('[install_callback] Failed to publish message:', error);
+			return null;
+		});
+		if (result == null) {
+			res.status(500).json({ error: 'Internal Server Error' });
+			return;
+		}
+	} else {
+		console.error('TOPIC_NAME not set');
+		res.status(400).json({ error: 'Failed to publish message'});
```
Comment (tapishr) [inline pages/api/github/callbacks/install.ts:24]: Add return in this else block

File: pages/api/github/callbacks/install.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+		const data = {
+			'repository_provider': 'github',
+			'installation_code': req.query.code
+		};
+		const msgType = 'install_callback';
+		console.info("Recieved installation code for github, published to ", topicName);
+		const result: string | null = await publishMessage(topicName, data, msgType)
+		.catch((error) => {
+			console.error('[install_callback] Failed to publish message:', error);
+			return null;
```
Comment (tapishr) [inline pages/api/github/callbacks/install.ts:16]: No need to return null, just return will do

File: pages/api/github/callbacks/install.ts (hunk 10 above)
Comment (tapishr) [reply pages/api/github/callbacks/install.ts:24]: The correct way to write this would be checking `if(!topicName)` instead of `if(topicName)`

Comment (MuskanPaliwal) [reply pages/api/github/callbacks/install.ts:8]: done

**Review by tapishr — APPROVED** [review]

**Pull request merged**